.github/workflows/publish.yml

```diff
@@ -1,21 +1,51 @@
-name: publish
+name: Build and Publish Docker Image
 
 on:
   push:
-    branches: [master, develop]
+    branches:
+        - "master"
+        - "develop"
+    tags:
+      - "v*.*.*"
 
 jobs:
-  publish-docker-image:
+  docker:
     runs-on: ubuntu-latest
     steps:
-      - name: Checkout the repository
+      -
+        name: Checkout
         uses: actions/checkout@v3
-
-      - name: Build and publish the Docker image
-        run: |
-          echo $GITHUB_TOKEN | docker login ghcr.io -u center-for-threat-informed-defense --password-stdin
-          docker build . --tag ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:$IMAGE_TAG
-          docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:$IMAGE_TAG
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          IMAGE_TAG: "${{ github.ref == 'refs/head/master' && 'latest' || 'develop' }}"
+      -
+        name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: |
+            ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api
+          tags: |
+            # set latest tag for master branch
+            type=raw,value=latest,enable=${{ github.ref == format('refs/heads/{0}', 'master') }}
+            # set develop tag for develop branch
+            type=ref,event=branch,enable=${{ github.ref == format('refs/heads/{0}', 'develop') }}
+            # set semver tag (vX.Y.Z) on git tag event
+            type=semver,pattern=v{{major}}.{{minor}}.{{patch}}
+            # set git short commit as Docker tag (e.g., sha-ad132f5)
+            type=sha
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      -
+        name: Login to GHCR
+        if: github.event_name != 'pull_request'
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      -
+        name: Build and push
+        uses: docker/build-push-action@v4
+        with:
+          context: .
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
```

Comment:
This is a fancier way of running:
```
docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:$IMAGE_TAG
```
except that we don't use `$IMAGE_TAG`. Instead we reference the Docker tags that were produced by the `metadata` action via `steps.meta.output.tags`. 

For example, a master merge event might trigger:
- `docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:sha-1234`
- `docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:latest`

A `develop` branch merge event might trigger:
- `docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:sha-2345`
- `docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:develop`

A `git tag v1.3.1` event might trigger:
- `docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:sha-3456`
- `docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:v1.3.1`